tools/SourceKit/tools/sourcekitd/lib/API/sourcekitdAPI-InProc.cpp

```diff
@@ -0,0 +1,969 @@
+//===--- sourcekitdAPI-InProc.cpp -----------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "DictionaryKeys.h"
+#include "sourcekitd/sourcekitd.h"
+#include "sourcekitd/Internal.h"
+#include "sourcekitd/CodeCompletionResultsArray.h"
+#include "sourcekitd/DocSupportAnnotationArray.h"
+#include "sourcekitd/TokenAnnotationsArray.h"
+#include "sourcekitd/Logging.h"
+#include "SourceKit/Support/UIdent.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/MemoryBuffer.h"
+
+#include <map>
+
+using namespace SourceKit;
+using namespace sourcekitd;
+using llvm::ArrayRef;
+using llvm::StringRef;
+using llvm::raw_ostream;
+
+namespace {
+class SKDDictionary;
+class SKDArray;
+class SKDCustomData;
+class SKDError;
+
+class SKDObject {
+public:
+  SKDObject(): RetainCount(0) {}
+  virtual ~SKDObject() {}
+
+  SKDObject(SKDObject const&) = delete;
+  SKDObject &operator=(SKDObject const&) = delete;
+
+  SKDObject *retain() {
+    ++RetainCount;
+    return this;
+  }
+
+  void release() {
+    if (RetainCount == 0) {
+      delete this;
+    } else {
+      --RetainCount;
+    }
+  }
+
+  virtual sourcekitd_variant_type_t getVariantType() = 0;
+
+  virtual void set(sourcekitd_uid_t Key, SKDObject *Value) {}
+  virtual void set(size_t Index, SKDObject *Value) {}
+
+  virtual SKDObject *get(sourcekitd_uid_t Key) { return nullptr; }
+
+  virtual SKDObject *get(size_t Index) { return nullptr; }
+  virtual size_t getCount() { return 0; }
+
+  virtual sourcekitd_uid_t getUID() { return nullptr; }
+  virtual Optional<int64_t> getInt64() { return Optional<int64_t>(); }
+  virtual Optional<StringRef> getString() { return Optional<StringRef>(); }
+  virtual bool getBool() { return false; }
+  virtual SKDDictionary *asDictionary() { return nullptr; }
+  virtual SKDArray *asArray() { return nullptr; }
+  virtual SKDCustomData *asCustomData() { return nullptr; }
+  virtual SKDError *asError() { return nullptr; }
+private:
+  size_t RetainCount;
+};
+
+class SKDDictionary: public SKDObject {
+public:
+  SKDDictionary() {}
+  SKDDictionary(const sourcekitd_uid_t *Keys,
+                const sourcekitd_object_t *Values,
+                size_t Count) {
```

Comment:
I recommend that for the APIs of the C++ objects always use `IntrusiveRefCntPtr`, don't use direct pointers to `sourcekitd_object_t` or `SKDObject`. We can convert between `sourcekitd_object_t` and an `IntrusiveRefCntPtr<SKDObject>` at the boundaries between the C API and the C++ one (inside the C functions implementations).